f600233d1b154d507b782f84604bcb12/update/sys_hub_action_type_definition_82657a551b2111107c458623604bcb46.xml

```diff
@@ -138,252 +138,324 @@
+            item.classification_instances.forEach(function (instance) {
+
+                // Check if need for entire task order duration is false
+                // If false, use the selected_periods array of the classification instance
+                if (!instance.need_for_entire_task_order_duration) {
+
+                    // For each period in the selected_periods array of the classification instance
+                    instance.selected_periods.forEach(function (period) {
+
+                        // Map period_type and base_period of the periods_estimate array
+                        var formattedPeriodOutput = periods_estimate.map(function (input) {
+                            return input.period.periodType + input.period.optionOrder;
+                        });
+
+                        var indexOfCurrentPeriod = formattedPeriodOutput.indexOf(period.periodType + period.optionOrder);
+                        // Insert into periods_estimate, under the specified period and period_line_items
+                        periods_estimate[indexOfCurrentPeriod].periodLineItems.push({
+                            "clin": "x001",
+                            "idiqClin": getClinFromDowTask(instance.dow_task_number, outputs.new_clins),
+                            "dowTaskNumber": instance.dow_task_number,
+                            "serviceOffering": item.service_offering.name,
+                            "itemDescriptionOrConfigSummary": instance.usage_description,
+                            "monthlyPrice": instance.monthly_price,
+                            "monthsInPeriod": monthsInPeriod(period.periodUnit, period.periodUnitCount),
+
+                        });
+                    });
+                }
+                // If instance is needed for the entire task_order_duration, add it to all period_line_items
+                else if (instance.need_for_entire_task_order_duration) {
+                    periods_estimate.forEach(function (period) {
+                        period.periodLineItems.push({
+                            "clin": "x001",
+                            "idiqClin": getClinFromDowTask(instance.dow_task_number, outputs.new_clins),
+                            "dowTaskNumber": instance.dow_task_number,
+                            "serviceOffering": item.service_offering.name,
+                            "itemDescriptionOrConfigSummary": instance.usage_description,
+                            "monthlyPrice": instance.monthly_price,
+                            // when entire period need to go in another layer to get period info
+                            "monthsInPeriod": monthsInPeriod(period.period.periodUnit, period.period.periodUnitCount),
+                        });
+                    });
+                }
+            });
+        }
+        // Else if the item includes estimated_environment_instances
+        else if (item.estimated_environment_instances) {
```

Comment:
Not that we can actually do proper unit tests...but in theory if we could, we'd probably want some of these big sections in their own functions.

Reply:
I agree it is basically the `selected_service_offerings` forEach loop adapted to work with `environment_instance` however this repetition is necessary due to Current Environments and Selected Service Offerings having different data structures. 

If we could do clean queries with joins for all the tables connected to Acquisition Package, the IGCE and DOW gathering actions wouldn't be so bloated.